internal/pkg/bulk/engine.go

```diff
@@ -98,13 +114,119 @@ func NewBulker(es esapi.Transport, tracer *apm.Tracer, opts ...BulkOpt) *Bulker
+		blkPool:               sync.Pool{New: poolFunc},
+		apikeyLimit:           semaphore.NewWeighted(int64(bopts.apikeyMaxParallel)),
+		tracer:                tracer,
+		remoteOutputConfigMap: make(map[string]map[string]interface{}),
+		bulkerMap:             make(map[string]Bulk),
+		remoteOutputErrorMap:  make(map[string]string),
+	}
+}
+
+func (b *Bulker) GetRemoteOutputErrorMap() map[string]string {
+	return b.remoteOutputErrorMap
+}
+
+func (b *Bulker) SetRemoteOutputError(name string, status string) {
+	b.remoteOutputErrorMap[name] = status
+}
+
+func (b *Bulker) GetBulker(outputName string) Bulk {
+	return b.bulkerMap[outputName]
+}
+
+func (b *Bulker) GetBulkerMap() map[string]Bulk {
+	return b.bulkerMap
+}
+
+func (b *Bulker) CancelFn() context.CancelFunc {
+	return b.cancelFn
+}
+
+func (b *Bulker) CreateAndGetBulker(zlog zerolog.Logger, outputName string, serviceToken string, outputMap map[string]map[string]interface{}) (Bulk, bool, error) {
+	hasConfigChanged := b.CheckRemoteOutputChanged(zlog, outputName, outputMap[outputName])
+	bulker := b.bulkerMap[outputName]
+	if bulker != nil && !hasConfigChanged {
+		return bulker, false, nil
+	}
+	if bulker != nil && hasConfigChanged {
+		cancelFn := bulker.CancelFn()
+		if cancelFn != nil {
+			cancelFn()
+		}
+	}
+	bulkCtx, bulkCancel := context.WithCancel(context.Background())
+	es, err := b.createRemoteEsClient(bulkCtx, outputName, serviceToken, outputMap)
+	if err != nil {
+		defer bulkCancel()
+		return nil, hasConfigChanged, err
+	}
+	// starting a new bulker to create/update API keys for remote ES output
+	newBulker := NewBulker(es, b.tracer)
+	newBulker.cancelFn = bulkCancel
+	b.bulkerMap[outputName] = newBulker
+
+	errCh := make(chan error)
+	go func() {
+		runFunc := func() (err error) {
+			zlog.Debug().Str("outputName", outputName).Msg("Bulker started")
+			return newBulker.Run(bulkCtx)
+		}
+
+		errCh <- runFunc()
+	}()
+	go func() {
+		select {
+		case err = <-errCh:
+		case <-bulkCtx.Done():
+			zlog.Debug().Str("outputName", outputName).Msg("Bulk context done")
+			err = bulkCtx.Err()
+		}
+	}()
+
+	return newBulker, hasConfigChanged, nil
+}
+
+func (b *Bulker) createRemoteEsClient(ctx context.Context, outputName string, serviceToken string, outputMap map[string]map[string]interface{}) (*elasticsearch.Client, error) {
+	hostsObj := outputMap[outputName]["hosts"]
+	hosts, ok := hostsObj.([]interface{})
+	if !ok {
+		return nil, fmt.Errorf("failed to get hosts from output: %v", hostsObj)
+	}
+	hostsStrings := make([]string, len(hosts))
+	for i, host := range hosts {
+		hostsStrings[i], ok = host.(string)
+		if !ok {
+			return nil, fmt.Errorf("failed to get hosts from output: %v", host)
+		}
+	}
+
+	cfg := config.Config{
+		Output: config.Output{
+			Elasticsearch: config.Elasticsearch{
+				Hosts:        hostsStrings,
+				ServiceToken: serviceToken,
+			},
+		},
 	}
+	es, err := es.NewClient(ctx, &cfg, false, elasticsearchOptions(
+		true, build.Info{},
```

Comment:
I think the `build.Info` here would be used to set user-agent headers with the fleet-server version number, is there an alternate way for us to set this here?

Reply:
Added the same `build.Info` as used in Fleet, do you know how to verify it? https://github.com/elastic/fleet-server/pull/3051/commits/10021a89ccdffcdcb63bc8107b60815d7042cc7b